systemd/shm_stop.service

```diff
@@ -1,7 +1,6 @@
 [Unit]
 Description=Shared memory cleanup process for Runtime
-After=dev_handler.service executor.service net_handler.service
```

Comment:
I think this is actually needed, or otherwise we need to remove it from the `multi-user.target` install. When you reboot the pi, you will see that `shm_stop` will be run with will probably cause some weird race conditions with `shm_start`.

Reply:
You're right about this, but do you think we could just add a `ConflictsWith=shm_start.service` and that would solve the issue? Let me try both, it might be that this works and the only reason why it didn't work before was the typo on the next line, for `OnFailure=` ...

Also about the SIGHUP, SIGINT, SIGTERM, SIGPIPE thing, what did you send those signals to? One of the Runtime processes? I thought that would cause said process to fail and `systemd` to restart the whole thing, no?

Reply:
Hmm ok it looks like both your solution and my solution don't work lol, I think what's happening is all of the services are running through once before it's restarted again and stable. I can't get `shm_stop.service` status to not say it failed at least once.

Reply:
Ok I think I figured it out, listing `shm_stop.service` as a conflicting service in `net_handler.service`, `dev_handler.service`, and `executor.service` prevents it from spawning. Here's the `systemctl status *.service` readout from the `runtime/systemd` directory, where you can see that `shm_start` started and exited properly, followed by the three main Runtime services (which are active now), and `shm_stop` never ran. You can tell that the three main processes started after `shm_start` because their PIDs are higher than `shm_start`!

```
● dev_handler.service - Device handler process for Runtime
   Loaded: loaded (/etc/systemd/system/dev_handler.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2021-01-04 13:31:41 PST; 40s ago
 Main PID: 389 (dev_handler)
    Tasks: 1 (limit: 4035)
   Memory: 448.0K
   CGroup: /system.slice/dev_handler.service
           └─389 /home/pi/runtime/dev_handler/dev_handler

Jan 04 13:31:41 raspberrypi systemd[1]: Started Device handler process for Runtime.
Jan 04 13:31:41 raspberrypi dev_handler[389]: INFO @ DEV_HANDLER        (Mon Jan  4 13:31:41 2021) DEV_HANDLER initialized.
Jan 04 13:31:41 raspberrypi dev_handler[389]: DEBUG @ DEV_HANDLER        (Mon Jan  4 13:31:41 2021) Polling now for /dev/ttyACM*.

● executor.service - Executor process for Runtime
   Loaded: loaded (/etc/systemd/system/executor.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2021-01-04 13:31:41 PST; 40s ago
 Main PID: 390 (executor)
    Tasks: 1 (limit: 4035)
   Memory: 1.7M
   CGroup: /system.slice/executor.service
           └─390 /home/pi/runtime/executor/executor

Jan 04 13:31:41 raspberrypi systemd[1]: Started Executor process for Runtime.
Jan 04 13:31:41 raspberrypi executor[390]: INFO @ EXECUTOR        (Mon Jan  4 13:31:41 2021) Executor initialized

● net_handler.service - Net handler process for Runtime
   Loaded: loaded (/etc/systemd/system/net_handler.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2021-01-04 13:31:41 PST; 40s ago
 Main PID: 392 (net_handler)
    Tasks: 3 (limit: 4035)
   Memory: 352.0K
   CGroup: /system.slice/net_handler.service
           └─392 /home/pi/runtime/net_handler/net_handler

Jan 04 13:31:41 raspberrypi systemd[1]: Started Net handler process for Runtime.
Jan 04 13:31:41 raspberrypi net_handler[392]: INFO @ NET_HANDLER        (Mon Jan  4 13:31:41 2021) Net handler initialized

● runtime_update.service - Update process for Runtime
   Loaded: loaded (/etc/systemd/system/runtime_update.service; disabled; vendor preset: enabled)
   Active: inactive (dead)

● shm_start.service - Shared memory create process for Runtime
   Loaded: loaded (/etc/systemd/system/shm_start.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Mon 2021-01-04 13:31:41 PST; 40s ago
  Process: 302 ExecStart=/home/pi/runtime/shm_wrapper/shm_start (code=exited, status=0/SUCCESS)
 Main PID: 302 (code=exited, status=0/SUCCESS)

Jan 04 13:31:41 raspberrypi systemd[1]: Starting Shared memory create process for Runtime...
Jan 04 13:31:41 raspberrypi shm_start[302]: INFO @ SHM        (Mon Jan  4 13:31:41 2021) SHM created
Jan 04 13:31:41 raspberrypi systemd[1]: shm_start.service: Succeeded.
Jan 04 13:31:41 raspberrypi systemd[1]: Started Shared memory create process for Runtime.

● shm_stop.service - Shared memory cleanup process for Runtime
   Loaded: loaded (/etc/systemd/system/shm_stop.service; enabled; vendor preset: enabled)
   Active: inactive (dead)
```

Reply:
And then here is the printout after killing one of the Runtime processes and it restarting:
```
● dev_handler.service - Device handler process for Runtime
   Loaded: loaded (/etc/systemd/system/dev_handler.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2021-01-04 13:32:53 PST; 8s ago
 Main PID: 591 (dev_handler)
    Tasks: 1 (limit: 4035)
   Memory: 120.0K
   CGroup: /system.slice/dev_handler.service
           └─591 /home/pi/runtime/dev_handler/dev_handler

Jan 04 13:32:53 raspberrypi systemd[1]: Started Device handler process for Runtime.
Jan 04 13:32:53 raspberrypi dev_handler[591]: INFO @ DEV_HANDLER        (Mon Jan  4 13:32:53 2021) DEV_HANDLER initialized.
Jan 04 13:32:53 raspberrypi dev_handler[591]: DEBUG @ DEV_HANDLER        (Mon Jan  4 13:32:53 2021) Polling now for /dev/ttyACM*.

● executor.service - Executor process for Runtime
   Loaded: loaded (/etc/systemd/system/executor.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2021-01-04 13:32:53 PST; 8s ago
 Main PID: 592 (executor)
    Tasks: 1 (limit: 4035)
   Memory: 628.0K
   CGroup: /system.slice/executor.service
           └─592 /home/pi/runtime/executor/executor

Jan 04 13:32:53 raspberrypi systemd[1]: Started Executor process for Runtime.
Jan 04 13:32:53 raspberrypi executor[592]: INFO @ EXECUTOR        (Mon Jan  4 13:32:53 2021) Executor initialized

● net_handler.service - Net handler process for Runtime
   Loaded: loaded (/etc/systemd/system/net_handler.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2021-01-04 13:32:53 PST; 8s ago
 Main PID: 593 (net_handler)
    Tasks: 3 (limit: 4035)
   Memory: 264.0K
   CGroup: /system.slice/net_handler.service
           └─593 /home/pi/runtime/net_handler/net_handler

Jan 04 13:32:53 raspberrypi systemd[1]: Started Net handler process for Runtime.
Jan 04 13:32:53 raspberrypi net_handler[593]: INFO @ NET_HANDLER        (Mon Jan  4 13:32:53 2021) Net handler initialized

● runtime_update.service - Update process for Runtime
   Loaded: loaded (/etc/systemd/system/runtime_update.service; disabled; vendor preset: enabled)
   Active: inactive (dead)

● shm_start.service - Shared memory create process for Runtime
   Loaded: loaded (/etc/systemd/system/shm_start.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Mon 2021-01-04 13:32:53 PST; 8s ago
  Process: 590 ExecStart=/home/pi/runtime/shm_wrapper/shm_start (code=exited, status=0/SUCCESS)
 Main PID: 590 (code=exited, status=0/SUCCESS)

Jan 04 13:32:53 raspberrypi systemd[1]: Starting Shared memory create process for Runtime...
Jan 04 13:32:53 raspberrypi shm_start[590]: INFO @ SHM        (Mon Jan  4 13:32:53 2021) SHM created
Jan 04 13:32:53 raspberrypi systemd[1]: shm_start.service: Succeeded.
Jan 04 13:32:53 raspberrypi systemd[1]: Started Shared memory create process for Runtime.

● shm_stop.service - Shared memory cleanup process for Runtime
   Loaded: loaded (/etc/systemd/system/shm_stop.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Mon 2021-01-04 13:32:53 PST; 8s ago
  Process: 589 ExecStart=/home/pi/runtime/shm_wrapper/shm_stop (code=exited, status=1/FAILURE)
 Main PID: 589 (code=exited, status=1/FAILURE)

Jan 04 13:32:53 raspberrypi systemd[1]: Starting Shared memory cleanup process for Runtime...
Jan 04 13:32:53 raspberrypi shm_stop[589]: INFO @ SHM        (Mon Jan  4 13:32:53 2021) SHM destroyed. RUNTIME FUNTIME HAD TOO MUCH FUN!!!
Jan 04 13:32:53 raspberrypi systemd[1]: shm_stop.service: Main process exited, code=exited, status=1/FAILURE
Jan 04 13:32:53 raspberrypi systemd[1]: shm_stop.service: Failed with result 'exit-code'.
Jan 04 13:32:53 raspberrypi systemd[1]: Failed to start Shared memory cleanup process for Runtime.
Jan 04 13:32:53 raspberrypi systemd[1]: shm_stop.service: Triggering OnFailure= dependencies.
```

Reply:
Can you explain what 'conflicts=' does exactly? From what I searched up, it'll stop the 'shm_stop.service` from running until the process that has this declared has stopped. So as soon as either `dev_handler`, `net_handler` or `executor` dies then the `shm_stop.service` will run. Is that right? Could you also add a comment or small addition to the wiki on `conflicts=` too please?

Reply:
Yes, will dp, I'll add a thing on the Wiki. But you got it basically right, `Conflicts=` basically declares that if the current unit is running, then no unit defined in `Conflicts=` can be running. So the reason I need to declare `shm_stop.service` in each of `dev_handler`, `net_handler`, and `executor` services as `Conflicts=` is to stop `shm_stop.service` from starting after `dev_handler`, `net_handler`, and `executor` services start running. It's not sufficient to declare in `shm_stop.service` that it goes after `dev_handler`, `net_handler`, and `executor`, or to declare in `shm_stop.service` that it is supposed to run on the failure of any of those three services, since `systemd` tries to boot up *all* enabled services on bootup, including `shm_stop.service`. Without the `Conflicts=` in the 3 main processes, `shm_start` would run first and exit, then all three main Runtime processes would start up, then `shm_stop` would run, clear the shared memory, and then start it all over again one more time and only then would `systemd` not run `shm_stop.service`. The `Conflicts=` ensures that `shm_stop` isn't started after the 3 main processes are brought up and running. Does that make sense?
@DanielMolina24

Reply:
Yea that makes sense! Thanks!